ant-bootstrap/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+            debug!("Removed {bootstrap_addr:?}");
+        }
+    }
+
+    pub fn sync(&mut self, other: &Self) {
+        for other_addr in other.0.iter() {
+            if let Some(bootstrap_addr) = self.get_addr_mut(&other_addr.addr) {
+                bootstrap_addr.sync(other_addr);
+            } else {
+                trace!(
+                    "Addr {:?} from other not found in self, inserting it.",
+                    other_addr.addr
+                );
+                self.insert_addr(other_addr);
+            }
+        }
+    }
+
+    pub fn update_addr_status(&mut self, addr: &Multiaddr, success: bool) {
+        if let Some(bootstrap_addr) = self.get_addr_mut(addr) {
+            bootstrap_addr.update_status(success);
+        } else {
+            debug!("Addr not found in cache to update, skipping: {addr:?}")
+        }
+    }
+}
+
+/// A addr that can be used for bootstrapping into the network
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct BootstrapAddr {
+    /// The multiaddress of the peer
+    pub addr: Multiaddr,
+    /// The number of successful connections to this address
+    pub success_count: u32,
+    /// The number of failed connection attempts to this address
+    pub failure_count: u32,
+    /// The last time this address was successfully contacted
+    pub last_seen: SystemTime,
+}
+
+impl BootstrapAddr {
+    pub fn new(addr: Multiaddr) -> Self {
+        Self {
+            addr,
+            success_count: 0,
+            failure_count: 0,
+            last_seen: SystemTime::now(),
+        }
+    }
+
+    pub fn peer_id(&self) -> Option<PeerId> {
+        multiaddr_get_peer_id(&self.addr)
+    }
+
+    pub fn update_status(&mut self, success: bool) {
+        if success {
+            if let Some(new_value) = self.success_count.checked_add(1) {
+                self.success_count = new_value;
+            } else {
+                self.success_count = 1;
+                self.failure_count = 0;
+            }
+        }
+        self.last_seen = SystemTime::now();
+        if !success {
+            if let Some(new_value) = self.failure_count.checked_add(1) {
+                self.failure_count = new_value;
+            } else {
+                self.failure_count = 1;
+                self.success_count = 0;
+            }
+        }
+    }
+
+    // An addr is considered reliable if it has more successes than failures
+    pub fn is_reliable(&self) -> bool {
+        self.success_count >= self.failure_count
+    }
+
+    /// Add the values from other into self.
+    pub fn sync(&mut self, other: &Self) {
+        trace!("Syncing our state {self:?} with and other: {other:?}.");
+        if self.last_seen == other.last_seen {
+            return;
+        }
+
+        self.success_count = self.success_count.saturating_add(other.success_count);
+        self.failure_count = self.failure_count.saturating_add(other.failure_count);
+
+        // if at max value, reset to 0
```

Comment:
shall be `capped`, not `reset` ?
i.e. no need to take this reset step

Reply:
This won't be an issue in the real world, but for some reason we reached the max for `success`, then only the `failures` will be tracked until it catches up with the `success`, making this peer useless as it will not be reliable.

So just wrap around to 0.